edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs

```diff
@@ -0,0 +1,735 @@
+namespace Microsoft.Azure.Devices.Edge.Hub.Http.Controllers
+{
+    using System;
+    using System.Net;
+    using System.Net.Http;
+    using System.Text;
+    using System.Threading.Tasks;
+    using System.Web;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class RegistryController : Controller
+    {
+        const string ContentType = "application/json; charset=utf-8";
+
+        readonly IRegistryApiClient apiClient;
+        readonly Task<IHttpRequestAuthenticator> authenticatorGetter;
+        readonly Task<IEdgeHub> edgeHubGetter;
+
+        public RegistryController(
+            IRegistryApiClient apiClient,
+            Task<IEdgeHub> edgeHubGetter,
+            Task<IHttpRequestAuthenticator> authenticatorGetter)
+        {
+            this.apiClient = Preconditions.CheckNotNull(apiClient, nameof(apiClient));
+            this.authenticatorGetter = Preconditions.CheckNotNull(authenticatorGetter, nameof(authenticatorGetter));
+            this.edgeHubGetter = Preconditions.CheckNotNull(edgeHubGetter, nameof(edgeHubGetter));
+        }
+
+        [HttpPut]
+        [Route("devices/{deviceId}/modules/{moduleId}")]
+        public async Task CreateOrUpdateModuleAsync(
+            [FromRoute] string deviceId,
+            [FromRoute] string moduleId,
+            [FromBody] Module module)
+        {
+            try
+            {
+                Events.ReceivedRequest(deviceId, moduleId);
+
+                try
+                {
+                    deviceId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(deviceId, nameof(deviceId)));
+                    moduleId = WebUtility.UrlDecode(Preconditions.CheckNonWhiteSpace(moduleId, nameof(moduleId)));
+                    Preconditions.CheckNotNull(module, nameof(module));
+
+                    if (!string.Equals(deviceId, module.DeviceId) || !string.Equals(moduleId, module.Id))
+                    {
+                        throw new ApplicationException("Device Id or module Id doesn't match between request URI and body.");
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Events.BadRequest(ex.Message);
+                    await this.SendResponseAsync(HttpStatusCode.BadRequest, this.FormatErrorResponseMessage(ex.Message));
+                    return;
+                }
+
+                IHttpRequestAuthenticator authenticator = await this.authenticatorGetter;
+                HttpAuthResult authResult = await authenticator.AuthenticateAsync(deviceId, Option.None<string>(), this.HttpContext);
+
+                if (!authResult.Authenticated)
+                {
+                    Events.AuthenticateFail(deviceId);
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
```

Comment:
output error message of the authResult? Applies to other instances too

Reply:
Current Registry API will return 401 status code with empty body.